django_mongodb/fields/auto.py

```diff
@@ -15,14 +16,27 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def get_prep_value(self, value):
-        # Override AutoField casting to integer.
-        return Field.get_prep_value(self, value)
+        if value is None:
+            return None
+        # Accept int for compatibility with Django's test suite which has many
+        # instances of manually assigned integer IDs, as well as for things
+        # like settings.SITE_ID which has a system check requiring an integer.
+        if isinstance(value, (ObjectId | int)):
+            return value
+        # If a digit doesn't match the length of an ObjectId, treat it as an
+        # integer. Manually assigned IDs won't be 24 digits long.
+        if isinstance(value, str) and value.isdigit() and len(value) != 24:
```

Comment:
I had to add a regression test to cover this condition:
```diff
diff --git a/tests/model_forms/tests.py b/tests/model_forms/tests.py
index d181585ca3..e5e99240d5 100644
--- a/tests/model_forms/tests.py
+++ b/tests/model_forms/tests.py
@@ -1511,8 +1511,9 @@ class ModelFormBasicTests(TestCase):
 <li>The URL: <input type="text" name="url" maxlength="40" required></li>""",
         )
 
-    def test_initial_values(self):
-        self.create_basic_data()
+    def test_initial_values(self, create_data=True):
+        if create_data:
+            self.create_basic_data()
         # Initial values can be provided for model forms
         f = ArticleForm(
             auto_id=False,
@@ -1623,6 +1624,26 @@ class ModelFormBasicTests(TestCase):
         test_art = Article.objects.get(id=art_id_1)
         self.assertEqual(test_art.headline, "Test headline")
 
+    def test_int_pks(self):
+        """
+        MongoAutoField supports numeric pks in ModelForm data, not just
+        ObjectId.
+        """
+        # The following lines repeat self.create_initial_data() but with
+        # manually assigned pks.
+        self.c1 = Category.objects.create(
+            pk=1, name="Entertainment", slug="entertainment", url="entertainment"
+        )
+        self.c2 = Category.objects.create(
+            pk=2, name="It's a test", slug="its-test", url="test"
+        )
+        self.c3 = Category.objects.create(
+            pk=3, name="Third test", slug="third-test", url="third"
+        )
+        self.w_royko = Writer.objects.create(name="Mike Royko", pk=1)
+        self.w_woodward = Writer.objects.create(name="Bob Woodward", pk=2)
+        self.test_initial_values(create_data=False)
+
     def test_m2m_initial_callable(self):
         """
         A callable can be provided as the initial value for an m2m field.
````

Reply:
I got confused by the [mongosh docs for ObjectId](https://www.mongodb.com/docs/manual/reference/method/ObjectId/) which says the constructor can take "The integer value, in seconds, is added to the Unix epoch to create the new timestamp." We don't need to do the "is `value` a numeric pk?" check until after `ObjectId()` (more common case), so yes, your code is what we end up with, except there is no need to call `ObjectId.is_valid()` which doubles the validation done by `ObjectId.__init__()`.